README
envbot - A modular IRC bot in bash
==================================
envbot is a modular IRC bot coded in bash.
https://launchpad.net/envbot

Features include:
 * SSL
 * IPv6
 * Transport (SSL, whatever) to server are also modules
 * Modularity
 * Loading, unloading and reloading of modules at runtime
 * Rehashing configuration at runtime
 * Advanced access control


Installing
----------
See the file INSTALL.


License
-------
envbot is licensed under GPL version 3, with the exceptions of:
	data/quotes.txt.example.pqf:
		The source (games-misc/fortune-mod-pqf/fortune-mod-pqf-6.0.ebuild from Gentoo portage)
		says it is under GPL-2.
	tools/bashdoc/bashdoc.sh
		GPL-2. bashdoc is a heavily updated and modified version of bashdoc from
		the sourcemage project.

Dependencies
------------
* bash - version 3.2.10 or later should work fine, but not tested on anything below 3.2.17
* Standard POSIX tools. Should be included on any recent and sane Linux
  distro.

Some transports and module have extra dependencies. Note that when it says
"you need the program" just having the library won't work. You actually need
a program with this name.
Transports:
	dev-tcp:
		The bash you use must support the pseudo device /dev/tcp. Debian is known
		to disable this. Most other distros are sane and have it on.
	netcat:
		You need the program netcat.
		This is for Debian users and others with a broken distro. If your
		distro supports it use dev-tcp transport instead.
		I have only tested with GNU netcat. (http://netcat.sourceforge.net/)
		Supports binding to a specific IP.
	gnutls:
		You need the program gnutls-cli. (http://www.gnutls.org/)
	openssl:
		You need the program openssl. (http://www.openssl.org/)
	socat:
		You need the program socat. (http://www.dest-unreach.org/socat/)
		Note that while socat support IPv4, IPv6, SSL and non-SSL it doesn't
		support both SSL and IPv6 at the same time if the version of socat
		is lower than 1.5.
		Supports binding to a specific IP.
Modules:
	sqlite3
		You need the program sqlite3. (http://www.sqlite.org/)
	factoids
		This depends on the sqlite3 module and therefore have the
		same dependencies as it
	seen
		This depends on the sqlite3 module and therefore have the
		same dependencies as it


Contributed modules
-------------------
These are extra modules in contrib/modules. They are not really
supported by the developers. Information about dependencies and
extra configuration options is in each contrib module.

If you want to use a contrib module the recommended way is:
  cd modules
  ln -s ../contrib/modules/m_modulename.sh
That way if the contrib module is updated you will get the new
version automatically.


Feedback
--------
We (the developers) would love to get feedback on what you like/dislike with
envbot, what features you want, and what you use it for.

Please also report any bugs you find at https://bugs.launchpad.net/envbot
(no login needed, but please enter your email so we can contact you if we
need more details about your problem).



ChangeLog.md
# ChangeLog

This is a overview of changes users may care about. Detailed changelog can be
acquired using git.

## 0.1-beta1

A lot is new in this release. Some of the highlights include:

 * Help command.
 * Coloured log output.
 * Modules that is split over several files in a subdirectory.
   See contrib/modules/m_perl for an example.
 * Lots of speed improvements by avoiding subshells when possible and such.
 * Checks in subshells that module loading will work before actually loading it.
   This is to avoid crashes on syntax errors in modules and such.
 * A new centralised bot command dispatching system that is easier to use
   than the old on_PRIVMSG and having every module parse it by itself.
 * API documentation is now auto generated from source code using bashdoc.
 * Extended and reworked internal API.
 * New more flexible module API.
 * Many new modules.
   * Core:
     * commands     - List commands from each module.
     * ctcp         - Responds to CTCPs send to the bot.
     * dice         - Roll a dice.
     * help         - Help command.
     * karma        - Karma module (the common ++/-- stuff)
     * nicktracking - tracks nicks <-> host mapping for channels the bot is in.
     * ping         - Provide some latency info and such.
     * uptime       - Tell the uptime of the bot.
   * Contrib (these either need extra deps or are not supported by the developers):
     * convert      - Convert between different units (needs GNU or BSD units installed).
     * perl         - Run perl code (not supported because it may not be safe).
 * Lots of bugfixes.

The rest of this change log entry is mainly for developers wanting to port their
code from 0.0.1 to 0.1:

 * 0.0.1 modules won't work without change, the both the module and core APIs
   have changed a lot. For example on_PRIVMSG should not be used now most of the
   time, instead see commands_register function.
   See doc/module_api2.txt and the auto generated API docs for details.
 * Quite a few Core API functions have changed parameter format:
   A lot now take "out parameters" instead of using $() construct. This is
   because subshells (caused by the $() construct) are slow.
 * Sadly one feature from trunk were too unstable to make it into 0.1, it will
   hopefully be in the next version. This feature was periodic events. This is
   the reason for some odd code related to transports status in this version
   ($transport_status and transport_alive to be specific).


# 0.0.1

Revisions | Description
--------- | -----------
475       | Updated man page.
472,474   | Fixed bash version check.
471       | Fixed an incorrect regular expression in lib/main.sh
470       | Made list_contains use grep -F instead.
469       | Fixed typo in ebuild.


# 0.0.1-rc1

Revisions | Description
--------- | -----------
465       | Backported fix to remove eval from various places.
463-464   | Fixed bug in modules/m_kick_ban.sh that made bot loose part of kick reason sometimes.
462       | Fixed bug in modules/m_join.sh that caused it to never send a reason on part.
459-461   | Added Gentoo ebuild for envbot.
457       | Made it work correctly on FreeBSD
456       | Fixed bug 29 (Channels not rejoined after ping timeout) and similar issue in modules/m_services.sh (ghost needed state not reset).
453       | Fixed for one aspect of bug 11.


# 0.0.1-beta5

Revisions | Description
--------- | -----------
450       | Merged r451 from trunk: Fix broken misc_clean_spaces (was broken after beta4).
443       | Made parse_hostmask_*() faster and added some missing quotes.
441       | Found that an INVITE hook was missing, added it.
430-435   | Added more numerics.
429       | Workaround for bug #21 added (Own nick desync during connect with ghost).


# 0.0.1-beta4

Revisions | Description
--------- | -----------
424       | Config version update: Added $config_log_raw. Defines if we should log raw lines or not (affects both STDOUT and logfiles)
420,423   | misc_clean_spaces was slow, changed to inline ways of stripping spaces.
416       | Got rid of deprecated parse_get_colon_arg function.
415       | Added some code to check for stuff that matches no hook, and found that there is no umode change hook because of this. Fixed.
411       | Made assignment of factoids work if there is more than one separator in string, yes this may be slower but it works.
410       | Fixed bug in factoids with forget command not reporting "I didn't have a factoid matching..." when none was found.


# 0.0.1-beta3

Revisions | Description
--------- | -----------
407       | Make error messages in transports more verbose.
402-406   | Make m_calc.sh check it's arguments on IRC better to make it secure.
401       | Fix bug that made modules_load break if a dependency failed to load.
398       | Made main envbot a wrapper script that sanitise the environment and then executes main bot.
395       | Some more functions are now internal in lib/log.sh, modules should use the log level versions.
393       | m_autojoin was slowing down rehash more than needed, fixed that.
392       | Fixed: Could not unload rehash, module_rehash_UNLOAD returned 127!
391       | Fixed bug that caused FINALISE hooks to never run
383       | Add bash version check (it turned out bash-3.1 and older was not supported).


# 0.0.1-beta2

Revisions | Description
--------- | -----------
379       | Actually more than one space between parameters on IRC is possible, yes that sucks, and no sane server would send it, but just to be on the safe side
377       | Added this ChangeLog.
376       | Fixed potential security problem in SQL cleaning code in module_sqlite3_clean_string (I can't find a way to abuse this so I don't think the risk is very high)
375       | Added access_log_action()
373-375   | Introduced more advanced logging API with several levels (info, warn, error and fatal)
371       | Optimised log_write() in lib/log.sh (something like 20 times as fast now if my testing is correct)


# 0.0.1-beta1


lib/numerics.sh
#!/bin/bash
# -*- coding: utf-8 -*-
###########################################################################
#                                                                         #
#  envbot - an IRC bot in bash                                            #
#  Copyright (C) 2007-2009  Arvid Norlander                               #
#                                                                         #
#  This program is free software: you can redistribute it and/or modify   #
#  it under the terms of the GNU General Public License as published by   #
#  the Free Software Foundation, either version 3 of the License, or      #
#  (at your option) any later version.                                    #
#                                                                         #
#  This program is distributed in the hope that it will be useful,        #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of         #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the          #
#  GNU General Public License for more details.                           #
#                                                                         #
#  You should have received a copy of the GNU General Public License      #
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.  #
#                                                                         #
###########################################################################

###########################################################################
#                                                                         #
# WARNING THIS FILE IS AUTOGENERATED. ANY CHANGES WILL BE OVERWRITTEN!    #
# See the source in tools/numerics.txt for comments about some numerics   #
# This file was generated with tools/build_numerics.sh                    #
#                                                                         #
###########################################################################
#---------------------------------------------------------------------
## Auto-generated list of numerics from tools/numerics.txt<br />
## This file contains a list of numerics that we currently use.
## It is therefore incomplete.<br />
## Because the list of variables in this file is so long, please see
## it's source for more details.
#---------------------------------------------------------------------

##########################
# Name -> number mapping #
##########################

numeric_RPL_WELCOME='001'
numeric_RPL_YOURHOST='002'
numeric_RPL_CREATED='003'
numeric_RPL_MYINFO='004'
numeric_RPL_ISUPPORT='005'
numeric_RPL_MAP='006'
numeric_RPL_MAPEND='007'
numeric_RPL_SNOMASK='008'
numeric_RPL_YOURID='042'
numeric_RPL_TRACEUSER='205'
numeric_RPL_STATSCLINE='213'
numeric_RPL_ENDOFSTATS='219'
numeric_RPL_UMODEIS='221'
numeric_RPL_STATSELINE='223'
numeric_RPL_RULES='232'
numeric_RPL_STATSUPTIME='242'
numeric_RPL_STATSCONN='250'
numeric_RPL_LUSERCLIENT='251'
numeric_RPL_LUSEROP='252'
numeric_RPL_LUSERUNKNOWN='253'
numeric_RPL_LUSERCHANNELS='254'
numeric_RPL_LUSERME='255'
numeric_RPL_ADMINME='256'
numeric_RPL_ADMINLOC1='257'
numeric_RPL_ADMINLOC2='258'
numeric_RPL_ADMINEMAIL='259'
numeric_RPL_TRYAGAIN='263'
numeric_RPL_LOCALUSERS='265'
numeric_RPL_GLOBALUSERS='266'
numeric_RPL_SILELIST='271'
numeric_RPL_ENDOFSILELIST='272'
numeric_RPL_AWAY='301'
numeric_RPL_USERHOST='302'
numeric_RPL_ISON='303'
numeric_RPL_TEXT='304'
numeric_RPL_UNAWAY='305'
numeric_RPL_UNAWAY='306'
numeric_RPL_WHOISREGNICK='307'
numeric_RPL_RULESSTART='308'
numeric_RPL_ENDOFRULES='309'
numeric_RPL_WHOISHELPOP='310'
numeric_RPL_WHOISUSER='311'
numeric_RPL_WHOISSERVER='312'
numeric_RPL_WHOISOPERATOR='313'
numeric_RPL_WHOWASUSER='314'
numeric_RPL_ENDOFWHO='315'
numeric_RPL_WHOISIDLE='317'
numeric_RPL_ENDOFWHOIS='318'
numeric_RPL_WHOISCHANNELS='319'
numeric_RPL_WHOISSPECIAL='320'
numeric_RPL_LISTSTART='321'
numeric_RPL_LIST='322'
numeric_RPL_LISTEND='323'
numeric_RPL_CHANNELMODEIS='324'
numeric_RPL_CHANNELURL='328'
numeric_RPL_CREATIONTIME='329'
numeric_RPL_WHOISACCOUNT='330'
numeric_RPL_NOTOPIC='331'
numeric_RPL_TOPIC='332'
numeric_RPL_TOPICWHOTIME='333'
numeric_RPL_USERIP='340'
numeric_RPL_INVITING='341'
numeric_RPL_INVITELIST='346'
numeric_RPL_ENDOFINVITELIST='347'
numeric_RPL_EXCEPTLIST='348'
numeric_RPL_ENDOFEXCEPTLIST='349'
numeric_RPL_VERSION='351'
numeric_RPL_WHOREPLY='352'
numeric_RPL_NAMREPLY='353'
numeric_RPL_LINKS='364'
numeric_RPL_ENDOFLINKS='365'
numeric_RPL_ENDOFNAMES='366'
numeric_RPL_BANLIST='367'
numeric_RPL_ENDOFBANLIST='368'
numeric_RPL_ENDOFWHOWAS='369'
numeric_RPL_INFO='371'
numeric_RPL_MOTD='372'
numeric_RPL_ENDOFINFO='374'
numeric_RPL_MOTDSTART='375'
numeric_RPL_ENDOFMOTD='376'
numeric_RPL_WHOISHOST='378'
numeric_RPL_YOUREOPER='381'
numeric_RPL_REHASHING='382'
numeric_RPL_TIME='391'
numeric_RPL_HOSTHIDDEN='396'
numeric_ERR_NOSUCHNICK='401'
numeric_ERR_NOSUCHSERVER='402'
numeric_ERR_NOSUCHCHANNEL='403'
numeric_ERR_CANNOTSENDTOCHAN='404'
numeric_ERR_TOOMANYCHANNELS='405'
numeric_ERR_WASNOSUCHNICK='406'
numeric_ERR_TOOMANYTARGETS='407'
numeric_ERR_NOTEXTTOSEND='412'
numeric_ERR_TOOMANYMATCHES='416'
numeric_ERR_UNKNOWNCOMMAND='421'
numeric_ERR_NOMOTD='422'
numeric_ERR_ERRONEUSNICKNAME='432'
numeric_ERR_NICKNAMEINUSE='433'
numeric_ERR_NICKTOOFAST='438'
numeric_ERR_TARGETTOOFAST='439'
numeric_ERR_USERNOTINCHANNEL='441'
numeric_ERR_NOTONCHANNEL='442'
numeric_ERR_USERONCHANNEL='443'
numeric_ERR_SUMMONDISABLED='445'
numeric_ERR_USERSDISABLED='446'
numeric_ERR_NONICKCHANGE='447'
numeric_ERR_NOTFORHALFOPS='460'
numeric_ERR_NEEDMOREPARAMS='461'
numeric_ERR_ALREADYREGISTERED='462'
numeric_ERR_ONLYSERVERSCANCHANGE='468'
numeric_ERR_LINKCHANNEL='470'
numeric_ERR_CHANNELISFULL='471'
numeric_ERR_UNKNOWNMODE='472'
numeric_ERR_INVITEONLYCHAN='473'
numeric_ERR_BANNEDFROMCHAN='474'
numeric_ERR_BADCHANNELKEY='475'
numeric_ERR_NEEDREGGEDNICK='477'
numeric_ERR_BANLISTFULL='478'
numeric_ERR_CANNOTKNOCK='480'
numeric_ERR_NOPRIVILEGES='481'
numeric_ERR_CHANOPRIVSNEEDED='482'
numeric_ERR_ATTACKDENY='484'
numeric_ERR_SECUREONLYCHAN='489'
numeric_ERR_ALLMUSTUSESSL='490'
numeric_ERR_NOOPERHOST='491'
numeric_ERR_NOREJOINONKICK='495'
numeric_ERR_CHANOWNPRIVNEEDED='499'
numeric_ERR_UMODEUNKNOWNFLAG='501'
numeric_ERR_USERSDONTMATCH='502'
numeric_RPL_LOGON='600'
numeric_RPL_LOGOFF='601'
numeric_RPL_WATCHOFF='602'
numeric_RPL_NOWON='604'
numeric_RPL_NOWOFF='605'
numeric_RPL_WATCHLIST='606'
numeric_RPL_ENDOFWATCHLIST='607'
numeric_RPL_WHOISSECURE='671'
numeric_RPL_MODULES='900'
numeric_RPL_ENDOFMODULES='901'
numeric_RPL_COMMANDS='902'
numeric_RPL_ENDOFCOMMANDS='903'
numeric_RPL_NOSPAM='931'
numeric_ERR_CENSORED='936'
numeric_ERR_ALREDYCENSORED='937'
numeric_ERR_NOTCENSORED='938'
numeric_ERR_SPAMFILTERLISTFULL='939'
numeric_RPL_ENDOFSPAMFILTER='940'
numeric_RPL_SPAMFILTER='941'
numeric_ERR_INVALIDNICK='942'
numeric_RPL_SILENCEREMOVED='950'
numeric_RPL_SILENCEADDED='951'
numeric_ERR_ALREADYSILENCE='952'
numeric_ERR_CANNOTDOCOMMAND='972'
numeric_ERR_CANNOTCHANGECHANMODE='974'

##########################
# Number -> name mapping #
##########################

numerics[1]='RPL_WELCOME'
numerics[2]='RPL_YOURHOST'
numerics[3]='RPL_CREATED'
numerics[4]='RPL_MYINFO'
numerics[5]='RPL_ISUPPORT'
numerics[6]='RPL_MAP'
numerics[7]='RPL_MAPEND'
numerics[8]='RPL_SNOMASK'
numerics[42]='RPL_YOURID'
numerics[205]='RPL_TRACEUSER'
numerics[213]='RPL_STATSCLINE'
numerics[219]='RPL_ENDOFSTATS'
numerics[221]='RPL_UMODEIS'
numerics[223]='RPL_STATSELINE'
numerics[232]='RPL_RULES'
numerics[242]='RPL_STATSUPTIME'
numerics[250]='RPL_STATSCONN'
numerics[251]='RPL_LUSERCLIENT'
numerics[252]='RPL_LUSEROP'
numerics[253]='RPL_LUSERUNKNOWN'
numerics[254]='RPL_LUSERCHANNELS'
numerics[255]='RPL_LUSERME'
numerics[256]='RPL_ADMINME'
numerics[257]='RPL_ADMINLOC1'
numerics[258]='RPL_ADMINLOC2'
numerics[259]='RPL_ADMINEMAIL'
numerics[263]='RPL_TRYAGAIN'
numerics[265]='RPL_LOCALUSERS'
numerics[266]='RPL_GLOBALUSERS'
numerics[271]='RPL_SILELIST'
numerics[272]='RPL_ENDOFSILELIST'
numerics[301]='RPL_AWAY'
numerics[302]='RPL_USERHOST'
numerics[303]='RPL_ISON'
numerics[304]='RPL_TEXT'
numerics[305]='RPL_UNAWAY'
numerics[306]='RPL_UNAWAY'
numerics[307]='RPL_WHOISREGNICK'
numerics[308]='RPL_RULESSTART'
numerics[309]='RPL_ENDOFRULES'
numerics[310]='RPL_WHOISHELPOP'
numerics[311]='RPL_WHOISUSER'
numerics[312]='RPL_WHOISSERVER'
numerics[313]='RPL_WHOISOPERATOR'
numerics[314]='RPL_WHOWASUSER'
numerics[315]='RPL_ENDOFWHO'
numerics[317]='RPL_WHOISIDLE'
numerics[318]='RPL_ENDOFWHOIS'
numerics[319]='RPL_WHOISCHANNELS'
numerics[320]='RPL_WHOISSPECIAL'
numerics[321]='RPL_LISTSTART'
numerics[322]='RPL_LIST'
numerics[323]='RPL_LISTEND'
numerics[324]='RPL_CHANNELMODEIS'
numerics[328]='RPL_CHANNELURL'
numerics[329]='RPL_CREATIONTIME'
numerics[330]='RPL_WHOISACCOUNT'
numerics[331]='RPL_NOTOPIC'
numerics[332]='RPL_TOPIC'
numerics[333]='RPL_TOPICWHOTIME'
numerics[340]='RPL_USERIP'
numerics[341]='RPL_INVITING'
numerics[346]='RPL_INVITELIST'
numerics[347]='RPL_ENDOFINVITELIST'
numerics[348]='RPL_EXCEPTLIST'
numerics[349]='RPL_ENDOFEXCEPTLIST'
numerics[351]='RPL_VERSION'
numerics[352]='RPL_WHOREPLY'
numerics[353]='RPL_NAMREPLY'
numerics[364]='RPL_LINKS'
numerics[365]='RPL_ENDOFLINKS'
numerics[366]='RPL_ENDOFNAMES'
numerics[367]='RPL_BANLIST'
numerics[368]='RPL_ENDOFBANLIST'
numerics[369]='RPL_ENDOFWHOWAS'
numerics[371]='RPL_INFO'
numerics[372]='RPL_MOTD'
numerics[374]='RPL_ENDOFINFO'
numerics[375]='RPL_MOTDSTART'
numerics[376]='RPL_ENDOFMOTD'
numerics[378]='RPL_WHOISHOST'
numerics[381]='RPL_YOUREOPER'
numerics[382]='RPL_REHASHING'
numerics[391]='RPL_TIME'
numerics[396]='RPL_HOSTHIDDEN'
numerics[401]='ERR_NOSUCHNICK'
numerics[402]='ERR_NOSUCHSERVER'
numerics[403]='ERR_NOSUCHCHANNEL'
numerics[404]='ERR_CANNOTSENDTOCHAN'
numerics[405]='ERR_TOOMANYCHANNELS'
numerics[406]='ERR_WASNOSUCHNICK'
numerics[407]='ERR_TOOMANYTARGETS'
numerics[412]='ERR_NOTEXTTOSEND'
numerics[416]='ERR_TOOMANYMATCHES'
numerics[421]='ERR_UNKNOWNCOMMAND'
numerics[422]='ERR_NOMOTD'
numerics[432]='ERR_ERRONEUSNICKNAME'
numerics[433]='ERR_NICKNAMEINUSE'
numerics[438]='ERR_NICKTOOFAST'
numerics[439]='ERR_TARGETTOOFAST'
numerics[441]='ERR_USERNOTINCHANNEL'
numerics[442]='ERR_NOTONCHANNEL'
numerics[443]='ERR_USERONCHANNEL'
numerics[445]='ERR_SUMMONDISABLED'
numerics[446]='ERR_USERSDISABLED'
numerics[447]='ERR_NONICKCHANGE'
numerics[460]='ERR_NOTFORHALFOPS'
numerics[461]='ERR_NEEDMOREPARAMS'
numerics[462]='ERR_ALREADYREGISTERED'
numerics[468]='ERR_ONLYSERVERSCANCHANGE'
numerics[470]='ERR_LINKCHANNEL'
numerics[471]='ERR_CHANNELISFULL'
numerics[472]='ERR_UNKNOWNMODE'
numerics[473]='ERR_INVITEONLYCHAN'
numerics[474]='ERR_BANNEDFROMCHAN'
numerics[475]='ERR_BADCHANNELKEY'
numerics[477]='ERR_NEEDREGGEDNICK'
numerics[478]='ERR_BANLISTFULL'
numerics[480]='ERR_CANNOTKNOCK'
numerics[481]='ERR_NOPRIVILEGES'
numerics[482]='ERR_CHANOPRIVSNEEDED'
numerics[484]='ERR_ATTACKDENY'
numerics[489]='ERR_SECUREONLYCHAN'
numerics[490]='ERR_ALLMUSTUSESSL'
numerics[491]='ERR_NOOPERHOST'
numerics[495]='ERR_NOREJOINONKICK'
numerics[499]='ERR_CHANOWNPRIVNEEDED'
numerics[501]='ERR_UMODEUNKNOWNFLAG'
numerics[502]='ERR_USERSDONTMATCH'
numerics[600]='RPL_LOGON'
numerics[601]='RPL_LOGOFF'
numerics[602]='RPL_WATCHOFF'
numerics[604]='RPL_NOWON'
numerics[605]='RPL_NOWOFF'
numerics[606]='RPL_WATCHLIST'
numerics[607]='RPL_ENDOFWATCHLIST'
numerics[671]='RPL_WHOISSECURE'
numerics[900]='RPL_MODULES'
numerics[901]='RPL_ENDOFMODULES'
numerics[902]='RPL_COMMANDS'
numerics[903]='RPL_ENDOFCOMMANDS'
numerics[931]='RPL_NOSPAM'
numerics[936]='ERR_CENSORED'
numerics[937]='ERR_ALREDYCENSORED'
numerics[938]='ERR_NOTCENSORED'
numerics[939]='ERR_SPAMFILTERLISTFULL'
numerics[940]='RPL_ENDOFSPAMFILTER'
numerics[941]='RPL_SPAMFILTER'
numerics[942]='ERR_INVALIDNICK'
numerics[950]='RPL_SILENCEREMOVED'
numerics[951]='RPL_SILENCEADDED'
numerics[952]='ERR_ALREADYSILENCE'
numerics[972]='ERR_CANNOTDOCOMMAND'
numerics[974]='ERR_CANNOTCHANGECHANMODE'

# End of generated file.
